### PR TITLE
operator: Report cluster operator version as per the new standard

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
@smarterclayton ok i think this aligns with what we discussed now.  ptal.

/cc @dmage 

i'm going to try to add an e2e to configuration_test that updates RELEASE_VERSION and confirms we see clusteroperator report that version, but what's here is ready for review.

**Comment by bparees** [comment]
very basic clusteroperator version upgrade/reporting test added.

A better test would ensure we don't report the new version until the operand deployment is at the new version, but i think what's here is a sufficient starting point.

File: pkg/apis/imageregistry/v1/types.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ const (
 	// StorageIncompleteUploadCleanupEnabled denotes whethere or not the registry storage
 	// medium is configured to automatically cleanup incomplete uploads
 	StorageIncompleteUploadCleanupEnabled = "StorageIncompleteUploadCleanupEnabled"
+
+	// VersionAnnotation reflects the version of the registry that this deployment
+	// is running.
+	VersionAnnotation = "imageregistry.operator.openshift.io/version"
```
Comment (smarterclayton) [inline pkg/apis/imageregistry/v1/types.go]: I used “release.openshift.io/version” in the network operator.  Better to have a convention.

File: pkg/clusteroperator/clusteroperator.go [hunk 2]
```diff
@@ -97,22 +88,24 @@ func (s *StatusHandler) Update(condtype osapi.ClusterStatusConditionType, status
...
+			newVersions := []osapi.OperandVersion{
+				{
+					Name:    "operator",
+					Version: newVersion,
+				},
+			}
+			if !reflect.DeepEqual(state.Status.Versions, newVersions) {
+				state.Status.Versions = newVersions
+				modified = true
+			}
```
Comment (smarterclayton) [inline pkg/clusteroperator/clusteroperator.go]: You need to reset versions to empty if new version is unset.

Comment (smarterclayton) [reply pkg/clusteroperator/clusteroperator.go]: Hrm.  I would not pass empty string to mean “not available”

File: pkg/operator/status.go [hunk 3]
```diff
@@ -71,9 +82,16 @@ func (c *Controller) syncStatus(cr *imageregistryv1.Config, deploy *appsapi.Depl
...
 		operatorAvailableMsg = "Deployment has minimum availability"
+		setOperatorVersion = isDeploymentStatusAvailableAndUpdated(deploy)
 	}
 
-	err := c.clusterStatus.Update(osapi.OperatorAvailable, operatorAvailable, operatorAvailableMsg)
+	deploymentVersion := ""
+	// if the current deployment has achieved availability at the new level, set the operator reported
+	// version to whatever level the deployment has achieved.
+	if setOperatorVersion {
+		deploymentVersion = deploy.Annotations[imageregistryv1.VersionAnnotation]
```
Comment (smarterclayton) [inline pkg/operator/status.go]: Hrm... ok, I guess.  Seems a bit weird but I can see why.

File: pkg/apis/imageregistry/v1/types.go (hunk 1 above)
Comment (bparees) [reply pkg/apis/imageregistry/v1/types.go]: works for me, i'll update it in the morning.

**Comment by bparees** [comment]
@smarterclayton updated the annotation name.  My interpretation of your other comments was everything was ok if slightly odd, but let me know if i missed something.

**Comment by bparees** [comment]
> I'm a bit worried about the special case newVersion = "" (which means that the version is not changed). But it looks good to merge.

`newVersion=""` means "we are not in a position to publish what version level we are at (e.g. because we are unavailable)"

**Pull request merged**